### PR TITLE
Reland "[JIT] during freezing, cast optional bias to half if weight is half"

### PR DESCRIPTION
**Pull request by davidberard98**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #77617

Original PR: #77295

Original commit message:
On GPU, conv errors if not all its inputs have the same dtype.

In the case of autocasting during freezing, what we see is:
1) inputs to conv are casted to half
2) inputs to batchnorm are not casted, so many are still floats
3) we try to fold conv + batchnorm, by finding different weight and bias such that conv(input, new_weight, new_bias) is equivalent to the original conv -> batchnorm.

If conv previously had an optional bias, then during freezing we will temporarily create a zero-valued bias as a placeholder for conv_bias. We want to construct it to have the same dtype as the weight input to conv, to avoid errors on GPU.

Reland changes:
There's a memory leak from cuda caching allocator that is a side effect of this fix. The memory leak causes the test to fail, though for some reason it didn't fail on CI in the last PR. This skips the tests for now.

### CODE REVIEW

**Review by eellison — APPROVED** [review]

**Comment by jjsjann123** [comment]
> There's a memory leak from cuda caching allocator that is a side effect of this fix. The memory leak causes the test to fail, though for some reason it didn't fail on CI in the last PR. This skips the tests for now.

I took a look at the failing test.

I noticed that the reported caching allocator leak is about in the same magnitude as of parameters in the model. Looking at the frozen graph, we are converting properties to prim::Constant in the graph. Looks like disabling that resolves the caching allocator failure for me locally. --> commenting out this line: https://github.com/pytorch/pytorch/blob/21d91cf774f3520ac552e92ee5de0f473f940cdf/torch/csrc/jit/passes/freeze_module.cpp#L131

Not sure how that would only break with caching allocator :confused: 
cc'ing @ngimel
